src/app/utils/firestore.ts

```diff
@@ -0,0 +1,62 @@
+import { Course, Question } from "../types/db";
+import { DocumentData, QueryDocumentSnapshot } from "firebase/firestore";
+
+export const courseConverter = {
+  toFirestore(course: Course): DocumentData {
+    return {
+      name: course.name,
+      location: course.location,
+      professors: course.professors,
+      students: course.students,
+      TAs: course.TAs,
+      onDuty: course.onDuty,
+      tags: course.tags,
+      // questions: course.questions,
+    };
+  },
+  fromFirestore(snapshot: QueryDocumentSnapshot, options: any): Course {
+    const data = snapshot.data(options);
+    if (!data) {
+      throw new Error("No data found in snapshot");
+    }
+    return {
+      name: data.name,
+      location: data.location,
+      professors: data.professors,
+      students: data.students,
+      TAs: data.TAs,
+      onDuty: data.onDuty,
+      tags: data.tags,
+      // questions: data.questions,
+    };
+  },
+};
+
+export const questionConverter = {
+  toFirestore(question: Question): DocumentData {
+    return {
+      title: question.title,
+      description: question.description,
+      public: question.public,
+      state: question.state,
+      timestamp: question.timestamp,
+      group: question.group,
+      tags: question.tags,
+    };
+  },
+  fromFirestore(snapshot: QueryDocumentSnapshot, options: any): Question {
+    const data = snapshot.data(options);
+    if (!data) {
```

Comment:
When would there be a case where `data` is null? Maybe we can do

```ts
return { ... } as Question
```

Reply:
+1 I like that idea. 
Looks like `data` might be null if 1) the document doesn't exist or 2) the user is not authenticated/have the right perms
https://stackoverflow.com/questions/53607050/documentsnapshot-is-always-null-for-a-valid-document-reference